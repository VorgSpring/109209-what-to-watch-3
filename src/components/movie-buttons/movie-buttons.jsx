import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {withRouter, Link} from 'react-router-dom';
import {getAuthorizationStatusSelector} from '../../selectors/authorization/authorization';
import {sendChangeFavoriteList} from '../../operations/favorites/favorites';
import {RoutePaths} from '../../constants/route-paths';
import {getReviewPathLink} from '../../helpers/get-links/get-links';
import {
  FavoriteStatus,
  IconFavoriteButton
} from '../../constants/favorite';
import {getPlayerLink} from '../../helpers/get-links/get-links';

export class MovieButtonsComponent extends PureComponent {
  constructor(props) {
    super(props);

    this.handleChangeFavoriteMovie = this.handleChangeFavoriteMovie.bind(this);
  }

  get nextStatus() {
    return this.props.isFavorite
      ? FavoriteStatus.REMOTE
      : FavoriteStatus.ADD;
  }

  get iconFavoriteButton() {
    return this.props.isFavorite
      ? IconFavoriteButton.IN_LIST
      : IconFavoriteButton.ADD;
  }

  handleChangeFavoriteMovie() {
    const {
      filmId,
      onChangeFavoritesList,
      isAuthorizationRequired,
      history
    } = this.props;

    if (isAuthorizationRequired) {
      history.push(RoutePaths.SING_IN);
      return;
    }

    onChangeFavoritesList({
      id: filmId,
      status: this.nextStatus
    });
  }

  render() {
    const {
      filmId,
      isFull
    } = this.props;

    return (
      <div className='movie-card__buttons'>
        <Link
          to={getPlayerLink(filmId)}
          className='btn btn--play movie-card__button'
        >
          <svg viewBox='0 0 19 19' width='19' height='19'>
            <use xlinkHref='#play-s' />
          </svg>
          <span>Play</span>
        </Link>
        <button
          className='btn btn--list movie-card__button'
          type='button'
          onClick={this.handleChangeFavoriteMovie}
        >
          <svg viewBox='0 0 19 20' width='19' height='20'>
            <use xlinkHref={this.iconFavoriteButton} />
          </svg>
          <span>My list</span>
        </button>
        {isFull &&
          <Link to={getReviewPathLink(filmId)} className='btn movie-card__button'>
            Add review
          </Link>}
      </div>
    );
  }
}

MovieButtonsComponent.defaultProps = {
  isFull: false
};

MovieButtonsComponent.propTypes = {
  isFull: PropTypes.bool,
  filmId: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isAuthorizationRequired: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  onChangeFavoritesList: PropTypes.func.isRequired
};

const MovieButtonsWrapper = withRouter(MovieButtonsComponent);

const mapStateToProps = (state) => ({
  isAuthorizationRequired: getAuthorizationStatusSelector(state)
});

const mapDispatchToProps = (dispatch) => ({
  onChangeFavoritesList: (body) => dispatch(sendChangeFavoriteList(body))
});

export const MovieButtons = connect(
    mapStateToProps, mapDispatchToProps
)(MovieButtonsWrapper);
