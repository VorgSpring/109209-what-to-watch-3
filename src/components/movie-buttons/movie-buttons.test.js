import React from 'react';
import {BrowserRouter} from 'react-router-dom';
import {MovieButtonsComponent} from './movie-buttons.jsx';
import renderer from 'react-test-renderer';

describe(`MoviePoster`, () => {
  const history = {
    push: () => {}
  };

  it(`should renders correctly`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <MovieButtonsComponent
            filmId={1}
            isFavorite={false}
            isAuthorizationRequired
            history={history}
            onChangeFavoritesList={() => {}}
            onPlay={() => {}}
          />
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should renders correctly with favorite icon`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <MovieButtonsComponent
            filmId={1}
            isFavorite
            isAuthorizationRequired={false}
            history={history}
            onChangeFavoritesList={() => {}}
            onPlay={() => {}}
          />
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should renders correctly with link to review`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <MovieButtonsComponent
            filmId={1}
            isFavorite={false}
            isAuthorizationRequired={false}
            isFull
            history={history}
            onChangeFavoritesList={() => {}}
            onPlay={() => {}}
          />
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
