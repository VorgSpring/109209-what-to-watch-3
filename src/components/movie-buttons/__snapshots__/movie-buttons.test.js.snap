// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePoster should renders correctly 1`] = `
<div
  className="movie-card__buttons"
>
  <a
    className="btn btn--play movie-card__button"
    href="/player/1"
    onClick={[Function]}
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </a>
  <button
    className="btn btn--list movie-card__button"
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#add"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
</div>
`;

exports[`MoviePoster should renders correctly with favorite icon 1`] = `
<div
  className="movie-card__buttons"
>
  <a
    className="btn btn--play movie-card__button"
    href="/player/1"
    onClick={[Function]}
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </a>
  <button
    className="btn btn--list movie-card__button"
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#in-list"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
</div>
`;

exports[`MoviePoster should renders correctly with link to review 1`] = `
<div
  className="movie-card__buttons"
>
  <a
    className="btn btn--play movie-card__button"
    href="/player/1"
    onClick={[Function]}
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </a>
  <button
    className="btn btn--list movie-card__button"
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#add"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
  <a
    className="btn movie-card__button"
    href="/film/1/review"
    onClick={[Function]}
  >
    Add review
  </a>
</div>
`;
